### ISSUE TITLE
TensorFlow support for Apple Silicon (M1 Chips)

### ISSUE THREAD

**Issue by MintStudios** [tensorflow/tensorflow#44751]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version: 2.1+ (I don't know specifics)
- Are you willing to contribute it (Yes/No): No, not enough repository knowledge.

**Describe the feature and the current behavior/state.**
This is not a feature per se, but a question. How soon would TensorFlow be available for the Apple Silicon macs announced today with the M1 chips?

**Will this change the current api? How?**
Hmmm, I don't know. Probably not.

**Who will benefit with this feature?**
This will benefit everybody who buys an apple silicon mac right now, and in the future. This is a big change, and no one would buy intel macs after a few years.

**Any Other info.**
I know python and numpy are being personally helped by apple for this transition.

**Comment by timocafe** [comment]
Well from the presentation of yesterday, specially timing [33 minutes](https://www.apple.com/apple-events/november-2020/), a person speaks about tensorflow, nothing super specific but it could be. Does a TF dev. could give more informations  about support of neural engine ? 
From my experience the neural engine is totally close to dev., only native execution of coreml can use it.

**Comment by MintStudios** [comment]
It would be really cool if TF could utilize the 16-core neural engine **and** their 8-core GPU.

**Comment by timocafe** [comment]
There is nothing against, what you can do at least for inference is to transform your TF model to [Coreml](https://coremltools.readme.io/docs) using Apple convertor. Then you will be able to execute your model "efficiently" on apple Hardware (inference only).

**Comment by MintStudios** [comment]
Well, my main problem would be training a model, as it takes more than an hour (I think) for my current model. When I get my hands on a mac, I'll see if it runs ok on rosetta 2. 

I just listened to the part of the event where they mention TensorFlow:
 "It also makes Mac Mini great for developers, scientists, and engineers. Utilizing deep learning technologies, like TensorFlow or Create ML, which are now accelerated by M1."
They say Mac Mini, but I'm sure the same thing goes for the MacBook Pro and Air.

**Comment by timocafe** [comment]
my 2 cents, do not train with an emulator, it is a terrible idea. And definitively a Mac is not the optimal platform for training.

**Comment by thecodacus** [comment]
> Well, my main problem would be training a model, as it takes more than an hour (I think) for my current model. When I get my hands on a mac, I'll see if it runs ok on rosetta 2.

I dont think we will need rosetta 2, as mac OS itself is a Unix machine so all the programming language and the compilers are natively running on mac for its core system to work. if we compile the TF library once(that might be bumpy), we should be good to go. what i wondering if we run the TF for training or inference(natively with TF not after porting to CoreML) will it utilize neural engine or the ML accelerators. maybe not, for that to work the processors needs to know that its a ML job and we are not specifically telling it that its a ML job. Not sure if its intelligent enough to do that on its own.

**Comment by MintStudios** [comment]
I really have no idea how python packaging works, but if something is done with Xcode, the new version should be able to compile for ARM if it's out yet. I might be wrong and maybe TF dependencies might be a problem.

**Comment by mihaimaruseac** [comment]
First question is if they meant TensorFlow or TensorFlow Lite or some other component of the ecosystem.

TensorFlow pips are only built for x86_64 at the moment.

**Comment by MintStudios** [comment]
Can TensorFlow pips be built for arm64  without any issue? If not, what is the main obstacle?
Hmmm, I'm not sure what TF component they meant. Probably TensorFlow.

**Comment by timocafe** [comment]
The main obstacle is: TF is build other backend (Eigen - CPU, CUDA- GPU, ...). It means you must be able to code all your layer, operators, etc  for these technologies. From my knowledge there is no TF  training for Apple chip (TF lite has something for inference Apple GPU).
Now if you compile from scratch TF on your Mac - M1. You will get a CPU version ARM (because Eigen works on ARM), but with no usage of Apple GPU or Neural Engine. Make a backend is a tough work, it needs a lot of effort, time, and working hour. Apple develop their own hardware/software I am not sure they care a lot about TF.

**Comment by MintStudios** [comment]
Hmmm. CPU training is OK for the time being, I'm not training huge models or anything. If you can, could you give me a super broad estimate of when the GPU or the neural engine could be used? Something like late 2020, mid 2021, etc. Or is this just going to come out of nowhere?

**Comment by MintStudios** [comment]
Maybe Apple would be of help somehow. Maybe.

**Comment by g4challenge** [comment]
There are several challenges involved with the architecture of the apple silicon M1 chip, the 4+4 cpu core architecture, the 8 GPU "cores" and the neural processing unit. What would be a great upside is the shared (but limited) memory.

CoreML is fine for inference of converted models, but full tensorflow would require a full backend which utilizes the full backend of components and does the memory management.

Inspiration and guidance could come from AMDs ROCm TF port.https://github.com/ROCmSoftwarePlatform/tensorflow-upstream. The build system is described here https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/rocm_docs/rocm-port-overview.md 

Another way would be to take the route of Metal or OpenCL - still the press event did not give much insight on how to actually use it. The issue for metal support is stale for some time #11085

**Comment by letalvoj** [comment]
> It would be really cool if TF could utilize the 16-core neural engine **and** their 8-core GPU.

The questions is whether ANE is even optimized for these kinds of workloads. 🤷‍♂️ According [to documentation](https://developer.apple.com/documentation/coreml/core_ml_api/personalizing_a_model_with_on-device_updates) there is just an option to "fine tune" (personalize) models locally.

**Comment by thecodacus** [comment]
I don't think we will see any new TF port for ANE, until Apple release any api thats similar to Cuda (only the API, as the implementation is Nvidia's proprietary tech), otherwise it will be very difficult to write the TF core logics to port to metal or existing ANE apis, great example is AMDs RCOM which is an api similar to CUDA. But I doubt Apple will come forward and make any effort, they will only try to push CoreML instead. And  also there are not too many people waiting for this feature (only those using macs most on them are using linux or have a separate linux machine). so most probably it will never happen from the open source community side as well. :(

**Comment by mihaimaruseac** [comment]
Note that we don't release `tensorflow-gpu` for macos anyway.

**Comment by krzysztofrusek** [comment]
I also wouldn't expect this to be supported out of the box quite soon.
Probably when MLIR and the new runtime get upstreamed, multiple backends will have better support.
An interesting alternative could be provided by Apple when auto diff will be part of swift, but this is just speculation.

**Comment by MintStudios** [comment]
So for now, it'll only work on the CPU. That's quite OK with me. If the CPU is better than Kaggle's, I'll be fine. I'm pretty new to deep learning, so I'm not doing heavy stuff. And maybe by the time I get farther into deep learning and ML, there might be support for the ANE and the GPU.

**Comment by vashat** [comment]
I wonder why Apple "announced" acceleration for Tensorflow on M1 macs on the Apple event. Are they outright lying, or did they mean something else? The specialy mentioned acceleration for Tensorflow and Create ML in the same sentence. And since Create ML is used for training I assumed that they also would provide acceleration for Tensorflow training.

**Comment by thecodacus** [comment]
>And since Create ML is used for training I assumed that they also would provide acceleration for Tensorflow training.

There in not development going on in the repo at the moment, that would only mean apple has compiled the packages for TF to support it, which is unlikely. Maybe they meant it is possible to do so, kind of like near future possibility. the only other explanation is that they just threw bunch of ML and DL terms to make it sound cool.

**Comment by masters3d** [comment]
> "Utilize ML frameworks like TensorFlow or Create ML, now accelerated by the M1 chip."  
![image](https://user-images.githubusercontent.com/6539412/99003546-169d6a00-24f3-11eb-9cfe-dce890edc25f.png)
from https://www.apple.com/newsroom/2020/11/introducing-the-next-generation-of-mac/ as of 2020-11-12

My only guess is that they are referring to Swift for Tensorflow but it would be nice to get clarification from someone who knows for sure. 
https://www.tensorflow.org/swift
https://github.com/tensorflow/swift

**Comment by dani0805** [comment]
I have opened a ticket on my apple developer account, lets see :-)

**Comment by dani0805** [comment]
The answer from developer support to my ticket: 

"Currently there is no documentation available for Tensorflow support on the M1 chip set. At the moment you can find general documentation for Apple Silicon in our documentation library.

I also forwarded your request as feedback to the appropriate team."

**Comment by dani0805** [comment]
So far the only way this could work is via

Tensorflow -> nGraph -> PlaidML -> Metal

No-one has tried though on arm, I have never bothered on my MBP as I don't have a GPU and I am doing all my TF testing on my Linux box. When I get my new MacBook I will give it a try.

**Comment by MintStudios** [comment]
Thanks @dani0805 for the ticket. I know apple doesn't care too much about stuff like this, but TensorFlow is huge and probably will be supported in the near future.

**Comment by yanina-libenson** [comment]
According to [this](https://www.apple.com/newsroom/2020/11/introducing-the-next-generation-of-mac/
) press release it's supported. --> "Utilize ML frameworks like TensorFlow or Create ML, now accelerated by the M1 chip."

**Comment by Degerz** [comment]
Right now what we can expect is that Tensorflow will be able to run on the CPUs in the M1 chip ... 

The biggest blocker to GPU acceleration is the Metal API. I've heard little to no indication that Apple rectified this issue on their end. CUDA and HIP (ROCm) support templatized kernels which are used everywhere in Tensorflow so Metal would need a similar capability to be even worth considering for porting the Eigen library. For the most part we should assume that Tensorflow won't support Apple GPUs for acceleration in the near future and likely in the distant future as well ...

With all that being said there might still be a way to optimize Tensorflow further for the M1's CPU by exploring the use of their AMX instruction set featured on some Apple processors ...

**Comment by robin7g** [comment]
TensorFlow on ARM is a thing. It works on NVIDIA Jetson Xaviers, which are ARM based. They use an ARM the CUDA library for GPU support. I am pretty sure Python will work on M1s which means likely TensorFlow CPU will work. I am very curious to know what "TensorFlow or Create ML, now accelerated by the M1 chip." is going to mean and if this needs to be driven by the community or if Apple is going to help.

**Comment by mihaimaruseac** [comment]
TF on ARM exists, but it's not part of the official releases on PyPI.

**Comment by msiracusa** [comment]
Accelerating TensorFlow Performance on Mac:
https://blog.tensorflow.org/2020/11/accelerating-tensorflow-performance-on-mac.html

**Comment by mihaimaruseac** [comment]
I guess that closes the issue then

**Comment by sumeshthakr** [comment]
I  am getting following error when I am importing tensor flow on MacBook Air m1
zsh: illegal hardware instruction  python

**Comment by mihaimaruseac** [comment]
The Apple TF on M1 chips is a private fork of TF owned by Apple. We don't have access to fixing code issues there

**Comment by kenthinson** [comment]
> Accelerating TensorFlow Performance on Mac:
> 
> https://blog.tensorflow.org/2020/11/accelerating-tensorflow-performance-on-mac.html
> 
> 
> 
> "TensorFlow users on Intel Macs or Macs powered by Apple’s new M1 chip can now take advantage of accelerated training using Apple’s Mac-optimized version of TensorFlow 2.4 and the new ML Compute framework."

I would not that the linked article only talks about CPU / GPU support. It doesn't mention the neural Engine. Perhaps performance is still being left on the table?￼

**Comment by dani0805** [comment]
> > Accelerating TensorFlow Performance on Mac:
> > https://blog.tensorflow.org/2020/11/accelerating-tensorflow-performance-on-mac.html
> > "TensorFlow users on Intel Macs or Macs powered by Apple’s new M1 chip can now take advantage of accelerated training using Apple’s Mac-optimized version of TensorFlow 2.4 and the new ML Compute framework."
> 
> I would not that the linked article only talks about CPU / GPU support. It doesn't mention the neural Engine. Perhaps performance is still being left on the table?￼

It should be possible to test this. I am quoting something that was written for iOS, but it should work on MacOS too https://github.com/hollance/neural-engine/blob/master/docs/is-model-using-ane.md

Core ML uses three different "engines" that are implemented in the private Espresso framework:

ANE — Espresso::ANERuntimeEngine
GPU — Espresso::MPSEngine and Espresso::MetalLowmemEngine
CPU — Espresso::BNNSEngine
It can switch between these engines when running your model. By looking at which engine name appears in the debugger's stack trace, you can see which engine / processor is currently being used.

To find out if the model (also) runs on the GPU or CPU, you can use the following symbolic breakpoints:

Espresso::MPSEngine::context::__launch_kernel
Espresso::BNNSEngine::convolution_kernel::__launch

**Comment by Tioz90** [comment]
> I am getting following error when I am importing tensor flow on MacBook Air m1
> zsh: illegal hardware instruction python

Same here

**Comment by MintStudios** [comment]
While on a [spreadsheet](https://docs.google.com/spreadsheets/d/1er-NivvuIheDmIKBVRu3S_BzA_lZT5z3Z-CxQZ-uPVs/edit) that shows apps/games that have support for apple silicon (native or translated), I found this:
https://docs.google.com/spreadsheets/d/1er-NivvuIheDmIKBVRu3S_BzA_lZT5z3Z-CxQZ-uPVs/edit
This has been released 10 days ago, and it's by Apple, and TensorFlow would be hardware accelerated.

**Comment by nyanye** [comment]
Although the apple's implementation does supports arm64 hardware acceleration there is no solution for some major thirdparty packages that works along with tensorflow such as matplotlib, opencv and etc. So using x86_64 tensorflow is still a required solution for specific usecases. Is there any possible way to use x86_64 tensorflow while using x86_64 python distribution on apple silicon mac? Thanks.

**Comment by TheMartian32** [comment]
So, by now, is there any form of support for Apple Silicon?

**Comment by Huibean** [comment]
Is this solved yet?

**Comment by TheMartian32** [comment]
@Huibean 

Yes. You'll need to install miniforge 3 and create a virtual env. The majority of the details are on Apple's TensorFlow fork. It's public so just type in apple TensorFlow Github.

**Comment by Huibean** [comment]
@TheMartian32 Thanks, just try that, but crash with another error when import TensorFlow
```
ImportError: Traceback (most recent call last):
  File "/Users/huibean/tensorflow_macos_venv/lib/python3.8/site-packages/tensorflow/python/pywrap_tensorflow.py", line 64, in <module>
    from tensorflow.python._pywrap_tensorflow_internal import *
ImportError: dlopen(/Users/huibean/tensorflow_macos_venv/lib/python3.8/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so, 6): no suitable image found.  Did find:
	/Users/huibean/tensorflow_macos_venv/lib/python3.8/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so: mach-o, but wrong architecture
	/Users/huibean/tensorflow_macos_venv/lib/python3.8/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so: mach-o, but wrong architecture
```

**Comment by TheMartian32** [comment]
@Huibean  Ah, sorry about that. This issue from that GitHub page seems to solve it: https://github.com/apple/tensorflow_macos/issues/3

Scroll down until it says Nov 18 and the comment should have ( at the time of writing this comment ) 13 likes and 5 hearts. If you encounter any other issues, feel free to tag me and I'd be happy to help.

**Comment by Huibean** [comment]
@TheMartian32 thx, but I am a little bit confused, this issue is created on  19 Nov 2020

**Comment by ptxmac** [comment]
Apple's closed source fork is not a good solution. For one it's missing the libraries (libtensorflow) and only provides the frameworks (which I assume have a subset of the functionality). Essentially the apple fork is only useful for running python with tensorflow, but any attempt at using the C interface will fail

**Comment by mihaimaruseac** [comment]
After #45404 you should be able to compile this repository on Apple M1 chips.

However, we still cannot provide full support as we lack devices to test on.

**Comment by Huibean** [comment]
@mihaimaruseac trying to build it today

**Comment by Huibean** [comment]
Now I can success import after build from source with master, thanks @mihaimaruseac

**Comment by mihaimaruseac** [comment]
Awesome. Thank you for confirming.

If you encounter bugs, please still open them here, though we cannot provide real support as for the other platforms at the moment due to lack of hardware and head count. But we'll try to fix things as it is possible.

**Comment by Huibean** [comment]
@mihaimaruseac I can import the module, trying to run some basic code, but it seems like an empty module
```
(base) ➜  Work ipython
Python 3.8.3 (default, Jul  2 2020, 11:26:31)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.16.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import tensorflow as tf

In [2]: tf.add(1, 2).numpy()
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-a78df43b9eac> in <module>
----> 1 tf.add(1, 2).numpy()

AttributeError: module 'tensorflow' has no attribute 'add'
```

**Comment by mihaimaruseac** [comment]
Two things:

1. Make sure you do the above from a different directory than the sources of TF when compiling from source
2. If the above doesn't work, can paste the output of `ls site-package/lib/tensorflow` (or where TF pip gets installed)?

**Comment by Huibean** [comment]
@mihaimaruseac 
I guess I don't have it in site-package
```
(base) ➜  site-packages pwd
/Users/huibean/anaconda3/lib/python3.8/site-packages
(base) ➜  site-packages ls tensor*
tensorboard_plugin_wit:
__init__.py          _vendor              wit_plugin_loader.py
__pycache__          static
_utils               wit_plugin.py

tensorboard_plugin_wit-1.8.0.dist-info:
INSTALLER        RECORD           entry_points.txt
METADATA         WHEEL            top_level.txt

tensorflow_estimator:
__init__.py __pycache__ _api        python

tensorflow_estimator-2.4.0.dist-info:
INSTALLER     METADATA      RECORD        WHEEL         top_level.txt
(base) ➜  site-packages
```

**Comment by mihaimaruseac** [comment]
What is the output of `pip list`?

**Comment by Huibean** [comment]
@mihaimaruseac I don't have it in pip
```
(base) ➜  tensorflow git:(master) pip list | grep tens
tensorboard-plugin-wit             1.8.0
tensorflow-estimator               2.4.0
typing-extensions                  3.7.4.2
widgetsnbextension                 3.5.1
```
I build the source with these steps, can't not get arm64 whl
```
 bazel build --config=macos_arm64 tensorflow/tools/pip_package:build_pip_package
./bazel-bin/tensorflow/tools/pip_package/build_pip_package /tmp/tensorflow_pkg
(base) ➜  tensorflow git:(master) ls /tmp/tensorflow_pkg
tensorflow-2.5.0-cp38-cp38-macosx_11_0_x86_64.whl
(base) ➜  tensorflow git:(master) pip install /tmp/tensorflow_pkg/tensorflow-2.5.0-cp38-cp38-macosx_11_0_x86_64.whl
ERROR: tensorflow-2.5.0-cp38-cp38-macosx_11_0_x86_64.whl is not a supported wheel on this platform.
```

**Comment by mihaimaruseac** [comment]
What is the output of `pip debug verbose`?

`pip list` seems to suggest the wheel was not installed, which makes sense. Now it's strange that `import tensorflow` works, but probably this is because of `tensorflow_estimator` depending on TF and faking it.

**Comment by webia1** [comment]
Can that help: https://github.com/apple/tensorflow_macos ?

**Comment by pratt-lightbeam** [comment]
any updates?

**Comment by aelmrabe** [comment]
Hi, is there any way to use the tensorflow-io in mac m1 ? Could I build the source code of tensorflow-io in the mac m1 machine ? The purpose is to use tensorflow-io with kafka in mac m1. Thanks.

**Comment by tux-o-matic** [comment]
@aelmrabe , just use Apple's own distribution of TensorFlow until it's merged into the main project https://github.com/apple/tensorflow_macos
